kubernetes/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -40,41 +41,78 @@ private[spark] class KubernetesClusterSchedulerBackend(
 
   val DEFAULT_NUMBER_EXECUTORS = 2
   val sparkExecutorName = s"spark-executor-${Random.alphanumeric take 5 mkString("")}".toLowerCase()
-  var executorPods = mutable.ArrayBuffer[String]()
+
+  // key is executor id, value is pod name
+  var executorToPod = mutable.Map.empty[String, String]
+  var executorID = 0
 
   val sparkDriverImage = sc.getConf.get("spark.kubernetes.driver.image")
   val clientJarUri = sc.getConf.get("spark.executor.jar")
   val ns = sc.getConf.get("spark.kubernetes.namespace")
+  val dynamicExecutors = Utils.isDynamicAllocationEnabled(sc.getConf)
+
+  // executor back-ends take their configuration this way
+  if (dynamicExecutors) {
+    sc.getConf.setExecutorEnv("spark.dynamicAllocation.enabled", "true")
+    sc.getConf.setExecutorEnv("spark.shuffle.service.enabled", "true")
+  }
 
-  override def start() {
+  override def start(): Unit = {
     super.start()
-    var i = 0
-    for(i <- 1 to getInitialTargetExecutorNumber(sc.conf)){
-      executorPods += createExecutorPod(i)
-    }
-    None
+    createExecutorPods(getInitialTargetExecutorNumber(sc.getConf))
   }
 
   override def stop(): Unit = {
-    for (i <- 0 to executorPods.length) {
-      client.pods().inNamespace(ns).withName(executorPods(i)).delete()
-    }
+    killExecutorPods(executorToPod.toVector)
     super.stop()
   }
 
   // Dynamic allocation interfaces
-  override def doRequestTotalExecutors(requestedTotal: Int): scala.concurrent.Future[Boolean] = {
-    return super.doRequestTotalExecutors(requestedTotal)
+  override def doRequestTotalExecutors(requestedTotal: Int): Future[Boolean] = {
+    logInfo(s"Received doRequestTotalExecutors: $requestedTotal")
+    val n = executorToPod.size
+    val delta = requestedTotal - n
+    if (delta > 0) {
+      logInfo(s"Adding $delta new executor Pods")
+      createExecutorPods(delta)
+    } else if (delta < 0) {
+      logInfo(s"Deleting ${-delta} new executor Pods")
```

Comment:
In this case, how do you plan to pick the executor pods to delete ? I can think of a few ways to do this, but ultimately the external shuffle service would be needed here to prevent losing shuffle blocks. (For safe deletion of executors). If this is in place, then perhaps we don't need to be particularly mindful of which executor pods get removed, so long as the resources are returned back to the cluster